prometheus/go_collector_go117.go

```diff
@@ -0,0 +1,321 @@
+
+//go:build go1.17
+// +build go1.17
+
+package prometheus
+
+import (
+	"math"
+	"runtime"
+	"runtime/metrics"
+	"strings"
+
+	//nolint:staticcheck // Ignore SA1019. Need to keep deprecated package for compatibility.
+	"github.com/golang/protobuf/proto"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/model"
+)
+
+type goCollector struct {
+	base baseGoCollector
+
+	// rm... fields all pertain to the runtime/metrics package.
+	rmSampleBuf []metrics.Sample
+	rmSampleMap map[string]*metrics.Sample
+	rmDescs     []*Desc
+	rmValTypes  []ValueType
+
+	// With Go 1.17, the runtime/metrics package was introduced.
+	// From that point on, metric names produced by the runtime/metrics
+	// package could be generated from runtime/metrics names. However,
+	// these differ from the old names for the same values.
+	//
+	// This field exist to export the same values under the old names
+	// as well.
+	msMetrics memStatsMetrics
+}
+
+// NewGoCollector is the obsolete version of collectors.NewGoCollector.
+// See there for documentation.
+//
+// Deprecated: Use collectors.NewGoCollector instead.
+func NewGoCollector() Collector {
+	descriptions := metrics.All()
+
+	// Generate a Desc and ValueType for each runtime/metrics metric.
+	descs := make([]*Desc, 0, len(descriptions))
+	valTypes := make([]ValueType, 0, len(descriptions))
+	sampleBuf := make([]metrics.Sample, 0, len(descriptions))
+	sampleMap := make(map[string]*metrics.Sample, len(descriptions))
+	for i := range descriptions {
+		d := &descriptions[i]
+		name, ok := nameFromRuntimeMetrics(d)
+		if !ok {
+			// TODO: Log this somewhere?
+			continue
+		}
+
+		// Set up sample buffer for reading, and a map
+		// for quick lookup of sample values.
+		sampleBuf = append(sampleBuf, metrics.Sample{
+			Name: d.Name,
+		})
+		sampleMap[d.Name] = &sampleBuf[len(sampleBuf)-1]
+
+		descs = append(descs, NewDesc(
+			name,
+			d.Description,
+			nil, nil,
+		))
+
+		valType := GaugeValue
+		if d.Cumulative {
+			valType = CounterValue
+		}
+		valTypes = append(valTypes, valType)
+	}
+	return &goCollector{
+		base:        newBaseGoCollector(),
+		rmSampleBuf: sampleBuf,
+		rmSampleMap: sampleMap,
+		rmDescs:     descs,
+		rmValTypes:  valTypes,
+		msMetrics:   goRuntimeMemStats(),
+	}
+}
+
+// Describe returns all descriptions of the collector.
+func (c *goCollector) Describe(ch chan<- *Desc) {
+	c.base.Describe(ch)
+	for _, i := range c.msMetrics {
+		ch <- i.desc
+	}
+	for _, desc := range c.rmDescs {
+		ch <- desc
+	}
+}
+
+// Collect returns the current state of all metrics of the collector.
+func (c *goCollector) Collect(ch chan<- Metric) {
+	// Collect base non-memory metrics.
+	c.base.Collect(ch)
+
+	// Populate runtime/metrics sample buffer.
+	metrics.Read(c.rmSampleBuf)
+
+	for i, sample := range c.rmSampleBuf {
+		desc := c.rmDescs[i]
+		switch sample.Value.Kind() {
+		case metrics.KindUint64:
+			ch <- MustNewConstMetric(desc, c.rmValTypes[i], float64(sample.Value.Uint64()))
```

Comment:
This will create lots of allocs from such simple functionality. I guess we can do that for now, but I would vote for separate function for new metrics, not in-place change 🤔 

Otherwise we could optimize it by making sure metrics is pre-build. Using constMetric for this sounds weird as it is... not constant at all.

Reply:
It's kind of similar to Kubernetes issue based on this https://github.com/prometheus/client_golang/discussions/917

Reply:
Got it. I'm not very familiar with Prometheus and what "const metric" really means. I was just doing the same thing the old code was doing. If there's a better way to do this I'm happy to switch to that instead.

Reply:
fixed all this to be more efficient. I now use Counter and Gauge for the new metrics, and the histogram type is no longer const.

I'm going to leave the Go 1.16 or earlier code to use the const API. this change is big enough as-is.

Reply:
Thanks 👍🏽